### PR TITLE
fix: Allow updates to scalars & other improvements

### PR DESCRIPTION
**Pull request by HCookie**
# Scalar updates

Allow the ScaleTensor of the loss function to have elements updated.

```python
ScaleTensor(test = (-1, torch.ones((2,))).update(test = torch.ones((2,)) * 3)

WeightedMSELoss().update_scalar('test', 3)
```

Additionally, add a `without` to allow subsetting without the given scalars.

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,7 @@ Keep it human-readable, your future self will thank you!
 ## [Unreleased](https://github.com/ecmwf/anemoi-training/compare/0.2.2...HEAD)
 
 ### Fixed
+- Allow updates to scalars
```
Comment (HCookie) [inline CHANGELOG.md:14]:
```suggestion
- Allow updates to scalars [#137](https://github.com/ecmwf/anemoi-training/pulls/137)
```

**Comment by sahahner** [comment]
What would be a use case for the `without` and `without_by_dim` functionalities that are provided by this PR?

**Comment by HCookie** [comment]
> What would be a use case for the `without` and `without_by_dim` functionalities that are provided by this PR?

I have just added a use case, 
In the verification metrics we can chose to remove scalars in the calculation.
It will also become useful in a later PR for stretched grids

File: src/anemoi/training/losses/weightedloss.py [hunk 2]
```diff
@@ -64,30 +65,45 @@ def __init__(
...
         """Scale a tensor by the variable_scaling.
 
         Parameters
         ----------
         x : torch.Tensor
             Tensor to be scaled, shape (bs, ensemble, lat*lon, n_outputs)
-        feature_indices:
-            feature indices (relative to full model output) of the features passed in pred and target
+        feature_indices: torch.Tensor, optional
+            feature indices (relative to full model output) of the features/variables passed in pred and target
```
Comment (sahahner) [inline src/anemoi/training/losses/weightedloss.py:86]: Should this be relative to the internal model output? I think we should follow the namespace of the data indices here.

Comment (sahahner) [reply src/anemoi/training/losses/weightedloss.py:86]: I noticed that for the validation metrics, "relative to the model output" is indeed correct. Either way, we should find a way to keep this description general.

Comment (HCookie) [reply src/anemoi/training/losses/weightedloss.py:86]: Agreed, we should keep the names consistent.
This is code transferred from the old losses, so not attached. I'll discuss with @Rilwan-Adewoyin who added this

Comment (HCookie) [reply src/anemoi/training/losses/weightedloss.py:86]: > to keep this description general.

Do you have a suggestion?

**Comment by sahahner** [comment]
>  In the verification metrics we can chose to remove scalars in the calculation. 

Is it also possible to apply this functionality on the training loss?

File: src/anemoi/training/losses/weightedloss.py (hunk 2 above)
Comment (sahahner) [reply src/anemoi/training/losses/weightedloss.py:86]: `feature indices (relative to model output to which this loss is applied) of the features/variables passed in pred and target`

**Comment by HCookie** [comment]
> Is it also possible to apply this functionality on the training loss?

Techically possible, yes, would we want that behaviour tho?

**Comment by sahahner** [comment]
> > Is it also possible to apply this functionality on the training loss?
> 
> Techically possible, yes, would we want that behaviour tho?

If it were not possible, I would suggest an assert. But if technically possible we can leave it that general.

File: src/anemoi/training/losses/weightedloss.py (hunk 2 above)
Comment (sahahner) [reply src/anemoi/training/losses/weightedloss.py:86]: I like the new name and docstring. We can update this one as well to `Indices to subset the calculated scalar with, by default None`.

**Review by sahahner — APPROVED** [review]
LGTM. Nice functionality that directly will be used in other PRs.
